plugins/SpectrumAnalyzer/SaControlsDialog.cpp

```diff
@@ -0,0 +1,227 @@
+/*
+ * SaControlsDialog.cpp - definition of SaControlsDialog class.
+ *
+ * Copyright (c) 2019 Martin Pavelek <he29/dot/HS/at/gmail/dot/com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SaControlsDialog.h"
+
+#include <QGridLayout>
+#include <QLabel>
+#include <QSizePolicy>
+#include <QSplitter>
+#include <QWidget>
+
+#include "ComboBox.h"
+#include "ComboBoxModel.h"
+#include "embed.h"
+#include "Engine.h"
+#include "LedCheckbox.h"
+#include "PixmapButton.h"
+#include "SaControls.h"
+#include "SaProcessor.h"
+
+
+// The entire GUI layout is built here.
+SaControlsDialog::SaControlsDialog(SaControls *controls, SaProcessor *processor) :
+	EffectControlDialog(controls),
+	m_controls(controls),
+	m_processor(processor)
+{
+	// Top level placement of sections is handled by QSplitter widget.
+	QHBoxLayout *master_layout = new QHBoxLayout;
+	QSplitter *display_splitter = new QSplitter(Qt::Vertical);
+	master_layout->addWidget(display_splitter);
+	master_layout->setContentsMargins(2, 6, 2, 8);
+	setLayout(master_layout);
+
+	// QSplitter top: configuration section
+	QWidget *config_widget = new QWidget;
+	QGridLayout *config_layout = new QGridLayout;
+	config_widget->setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Minimum);
+	config_widget->setMaximumHeight(m_configHeight);
+	config_widget->setLayout(config_layout);
+	display_splitter->addWidget(config_widget);
+
+	// Pre-compute target pixmap size based on on monitor DPI.
+	// Using setDevicePixelRatio() on pixmap allows the SVG image to be razor
+	// sharp on High-DPI screens, but the desired size must be manually
+	// enlarged. No idea how to make Qt do it in a more reasonable way.
+	float iconSize = 22.0 * devicePixelRatio();
+	float buttonSize = 1.2 * iconSize;
+
+	// pause and freeze buttons
```

Comment:
With the `PixmapButton`s, `LedCheckBox`es and `ComboBox`es, you're always doing the same. Especially for the pixmap buttons. It could increase readability if you'd write a lambda function which does it once and then call it for every button. E.g. like `auto initPixmapButton = [](/*parameters*/) { /*code*/ }; PixmapButton* pauseButton = initPixmapButton(/*...*/);`.

But it can also sometimes decrease readability :grimacing: , so consider this just as a hint.

Reply:
I think I prefer plain old code here. :)) At least now that it is finished -- it would have probably saved me some time during development, though. I'm not that familiar with lambda functions, so I mostly just use them when it is needed..

Reply:
Lambda function can be really fun and save you from writing "normal functions" and passing all the stuff to those :smile: 

For this case, I'm OK if you hit the resolve button.